pkg/discovery/dtofactory/service_entity_dto_builder.go

```diff
@@ -95,14 +98,81 @@ func (builder *ServiceEntityDTOBuilder) BuildDTOs() ([]*proto.EntityDTO, error)
 		//3. create it
 		entityDto, err := ebuilder.Create()
 		if err != nil {
-			glog.Errorf("failed to create service[%s] EntityDTO: %v", displayName, err)
+			glog.Errorf("Failed to create service[%s] EntityDTO: %v", displayName, err)
 			continue
 		}
 		glog.V(4).Infof("service DTO: %++v", entityDto)
 		result = append(result, entityDto)
 	}
 
-	return result, nil
+	return result
+}
+
+// Create sold commodities for Service entity.
+// Currently, only NumberReplicas is sold by Service
+func (builder *ServiceEntityDTOBuilder) createCommoditySold(
+	ebuilder *sdkbuilder.EntityDTOBuilder, pods []*api.Pod, serviceName string) error {
+	var controllers = make(map[string]*repository.K8sController)
+	readyPods := 0
+	for _, pod := range pods {
+		// Get the pod controller info from the podToController cache in ClusterScraper
+		ownerInfo, _, _, err := builder.ClusterScraper.GetPodControllerInfo(pod, true)
+		if err != nil || util.IsOwnerInfoEmpty(ownerInfo) {
+			// The pod does not have a controller
+			continue
+		}
+		if _, found := controllers[ownerInfo.Uid]; found {
+			// We've already visited this controller
+			if util.PodIsReady(pod) {
+				readyPods++
+			}
+			continue
+		}
+		// Get the cached controller from the controller cache in ClusterSummary
+		controller, found := builder.ClusterSummary.ControllerMap[ownerInfo.Uid]
+		if !found {
+			// No cached controller
+			continue
+		}
+		if controller.Replicas == nil || *controller.Replicas < 1 {
+			// No valid replicas
+			continue
+		}
+		controllers[controller.UID] = controller
+		if util.PodIsReady(pod) {
+			readyPods++
+		}
+	}
+	if len(controllers) == 0 {
+		glog.Errorf("No controllers are found for any pod that provides to service %v.", serviceName)
+		return nil
+	}
+	// The list of pods that provide to the service may belong to multiple controllers
+	// Determine capacity by getting the max of all replicas
+	var replicas int64
+	for _, controller := range controllers {
+		if *controller.Replicas > replicas {
+			replicas = *controller.Replicas
+		}
+	}
+	used := float64(readyPods)
+	capacity := float64(replicas)
```

Comment:
what will happen if the replica is 0?

Reply:
never mind, doesn't look like it will happen here.

Reply:
Right, controllers without replicas, or with replicas < 1 will not be considered.